### ISSUE TITLE
Http2: Hyper client gets stuck if too many requests are spawned

### ISSUE THREAD

**Issue by staticgc** [hyperium/hyper#2419]
I have a golang https server to which a rust client sends lot of parallel requests.
After few hundreds of requests, it stops and apparently there is no TCP communication.

The REST API accepts a byte buffer in a body and responds back with its length in a json.

To reproduce, the entire code is at: https://github.com/staticgc/hyper-stuck
It has golang server, Rust Client & Golang Client
The build instructions are very straightforward 

The default max number of futures spawned are [400](https://github.com/staticgc/hyper-stuck/blob/main/test.env)
This apparently is more than the max number of http2 streams at server (which is 250)

If the count is reduced to 200 it works.

Other observations:

1. Golang client works with high goroutine count
2. Increasing connections in Rust client from 1 to say 3 causes more requests to be processed but again gets stuck
3. With 400 futures/requests, when the number of connections = 10 (chosen arbitrarily) it worked ok.
4. When [HTEST_BUF_SIZE](https://github.com/staticgc/hyper-stuck/blob/main/test.env) is reduced to say 1KB then even with 1 connection & high future count it works
5. If the rust client stops, it does so after predictable number of requests

Number of connections here refers to http2 negotiated connection which hyper (I think) creates only 1 per `Client` instance.
So changing [HTEST_CONN_COUNT](https://github.com/staticgc/hyper-stuck/blob/main/test.env) changes the number of `Client` instances that are created.

Also to prevent initial flooding at the server, the rust client makes a single http2 request on each `Client` instance and then issues the parallel requests.

```rust
for i in 0..conn_count {
     send_req_https(client_vec[i].clone(), bufsz, url.as_str()).await?;
}
```

Able to reproduce on: Mac OS & Cent OS Linux

Edit: Apologies for some typos in build instructions in the repo above. I have fixed those. Let me know here if anything remains.

**Comment by staticgc** [comment]
I also changed the `#[tokio::main]` to use single threaded basic schedular and after it got stuck this was the call graph (on mac).

It seems it is waiting for event from OS (kevent) to move ahead.

```
Call graph:
    2865 Thread_8562679   DispatchQueue_1: com.apple.main-thread  (serial)
      2865 start  (in libdyld.dylib) + 1  [0x7fff61883015]
        2865 main  (in hyper-stuck) + 34  [0x10322a072]
          2865 std::rt::lang_start::h1a153dc93960f3ad  (in hyper-stuck) + 65  [0x10332fb01]  rt.rs:65
            2865 std::rt::lang_start_internal::h36ccce6e8a047133  (in hyper-stuck) + 644  [0x10364fc64]  panicking.rs:345
              2865 std::rt::lang_start::_$u7b$$u7b$closure$u7d$$u7d$::h3ca2bfef5797c7c8  (in hyper-stuck) + 20  [0x10332fb24]  rt.rs:66
                2865 std::sys_common::backtrace::__rust_begin_short_backtrace::h3bb497384788fb92  (in hyper-stuck) + 17  [0x103202761]  backtrace.rs:125
                  2865 core::ops::function::FnOnce::call_once::h5ebd9be2d6f4a542  (in hyper-stuck) + 14  [0x1032fd7ee]  function.rs:227
                    2865 hyper_stuck::main::hac5673c1b3a36a96  (in hyper-stuck) + 137  [0x103229fd9]  main.rs:133
                      2865 tokio::runtime::Runtime::block_on::h479291c746868d90  (in hyper-stuck) + 278  [0x10331ff56]  mod.rs:450
                        2865 tokio::runtime::basic_scheduler::BasicScheduler$LT$P$GT$::block_on::h30fcf5d82cf05d14  (in hyper-stuck) + 189  [0x10326531d]  basic_scheduler.rs:136
                          2865 tokio::runtime::basic_scheduler::InnerGuard$LT$P$GT$::block_on::h2bc6b4e076e10001  (in hyper-stuck) + 78  [0x103264fee]  basic_scheduler.rs:405
                            2865 tokio::runtime::basic_scheduler::Inner$LT$P$GT$::block_on::h8205f88ce1328514  (in hyper-stuck) + 29  [0x10326470d]  basic_scheduler.rs:176
                              2865 tokio::runtime::basic_scheduler::enter::h293894db779d083a  (in hyper-stuck) + 718  [0x1032659ae]  basic_scheduler.rs:266
                                2865 tokio::macros::scoped_tls::ScopedKey$LT$T$GT$::set::h56389ade64ea81f8  (in hyper-stuck) + 127  [0x1032a2d1f]  scoped_tls.rs:61
                                  2865 tokio::runtime::basic_scheduler::enter::_$u7b$$u7b$closure$u7d$$u7d$::h701802ea7bbbd133  (in hyper-stuck) + 52  [0x103265a84]  basic_scheduler.rs:266
                                    2865 tokio::runtime::basic_scheduler::Inner$LT$P$GT$::block_on::_$u7b$$u7b$closure$u7d$$u7d$::h199c0b1c5b3c4adc  (in hyper-stuck) + 1201  [0x103264bd1]  basic_scheduler.rs:211
                                      2865 _$LT$tokio..runtime..driver..Driver$u20$as$u20$tokio..park..Park$GT$::park::h6a77d4dd7ca55461  (in hyper-stuck) + 24  [0x103559e28]  driver.rs:196
                                        2865 _$LT$tokio..park..either..Either$LT$A$C$B$GT$$u20$as$u20$tokio..park..Park$GT$::park::h7f22102478fcba53  (in hyper-stuck) + 105  [0x10356ae59]  either.rs:30
                                          2865 _$LT$tokio..time..driver..Driver$LT$P$GT$$u20$as$u20$tokio..park..Park$GT$::park::h298260be2a66467d  (in hyper-stuck) + 29  [0x10359cdfd]  mod.rs:391
                                            2865 tokio::time::driver::Driver$LT$P$GT$::park_internal::h0c92f7c2c1414e75  (in hyper-stuck) + 855  [0x10359c0b7]  mod.rs:219
                                              2865 _$LT$tokio..park..either..Either$LT$A$C$B$GT$$u20$as$u20$tokio..park..Park$GT$::park_timeout::h6bb7bbd7924a5514  (in hyper-stuck) + 127  [0x10356afef]  either.rs:37
                                                2865 _$LT$tokio..process..imp..driver..Driver$u20$as$u20$tokio..park..Park$GT$::park_timeout::h336fe5900728c1c4  (in hyper-stuck) + 34  [0x10355e5b2]  driver.rs:86
                                                  2865 _$LT$tokio..signal..unix..driver..Driver$u20$as$u20$tokio..park..Park$GT$::park_timeout::h2803f5690a8d9dba  (in hyper-stuck) + 34  [0x1035ac102]  driver.rs:156
                                                    2865 _$LT$tokio..io..driver..Driver$u20$as$u20$tokio..park..Park$GT$::park_timeout::h7d280450da7be4f6  (in hyper-stuck) + 58  [0x10358b5fa]  mod.rs:238
                                                      2865 tokio::io::driver::Driver::turn::hab664d2964cd5899  (in hyper-stuck) + 323  [0x10358abf3]  mod.rs:162
                                                        2865 mio::poll::Poll::poll::hbea28d9af2243ed7  (in hyper-stuck) + 80  [0x103608510]  poll.rs:314
                                                          2865 mio::sys::unix::selector::kqueue::Selector::select::hfd9ef3fe90682fa2  (in hyper-stuck) + 327  [0x10360a307]  kqueue.rs:115
                                                            2865 kevent  (in libsystem_kernel.dylib) + 10  [0x7fff619d4bea]
```

**Comment by staticgc** [comment]
I think I might have found the issue.

When the number of active streams exceed 250 (which is the default limit), the future polled [here](https://github.com/hyperium/hyper/blob/dfa1bb291d71978820e8077be94c1d5837935bbc/src/proto/h2/client.rs#L203) returns `Poll::Pending`. Now the parent future i.e. `ClientTask` does not get woken up again when stream count goes below the threshold.

@seanmonstar

**Comment by staticgc** [comment]
Here is a work-around:

Wrap `hyper::client::Client` and guard the `request` method with a semaphore.

If more parallel requests are needed, create multiple `CustomClient`

```rust

use hyper::{Body, {client::HttpConnector}};
use hyper_openssl::HttpsConnector;
use tokio::sync::Semaphore;

#[derive(Clone)]
struct CustomClient {
    client: Client<HttpsConnector<HttpConnector>>,
    guard: Arc<Semaphore>,
}

impl CustomClient {
    fn new() -> Result<Self, Error> {
        let client = CustomClient::new_hyper_client()?;
        let guard = Arc::new(Semaphore::new(240)); //Harcoded limit just below threshold 250

        Ok(CustomClient{
            client,
            guard,
        })
    }

    async fn request(&self, req: Request<Body>) -> Result<Response<Body>, Error> {
        //Is allowed to make request? If not then wait.
        let _permit = self.guard.acquire().await?;

        let rsp = self.client.request(req).await?;
        Ok(rsp)
    }
}
```

**Comment by cynecx** [comment]
Ping. Somehow I came across this issue trying out some stuff between go and tokio/hyper/h2. Since it seems like an issue in h2, it would make sense to move the issue there?

**Comment by staticgc** [comment]
> Ping. Somehow I came across this issue trying out some stuff between go and tokio/hyper/h2. Since it seems like an issue with h2, it would make sense to move the issue there?

@seanmonstar

**Comment by silence-coding** [comment]
How's this going?

**Comment by fasterthanlime** [comment]
Here's a nice self-contained reproduction for this bug: https://github.com/fasterthanlime/h2-repro

**Comment by jfourie1** [comment]
FWIW I observed that an existing waker is silently dropped [here](https://github.com/hyperium/h2/blob/756e2524ac53c0cf9810efcba0b92171b7d80e42/src/proto/streams/stream.rs#L325). In other words the following assert would fail if added to the top of the function : `assert!(self.send_task.is_none())`

**Comment by jfourie1** [comment]
As a quick test / hack I added the following to the start of the function mentioned above : 
```rust
        if let Some(task) = self.send_task.take() {
            task.wake();
        }
```
This seems to fix the issue for me but still trying to figure out what is going on. Won't recommend just blindly applying this patch.

**Comment by jfourie1** [comment]
Some further analysis:

Looking at [this loop's](https://github.com/hyperium/hyper/blob/287d712483aec6671427438d60ed2a72f856fd9f/src/proto/h2/client.rs#L227) first iteration the following seems to be happening for the request that gets stuck:
The stream is not marked as pending yet so [this](https://github.com/hyperium/hyper/blob/287d712483aec6671427438d60ed2a72f856fd9f/src/proto/h2/client.rs#L227) returns `Ok(())`.
The stream is marked pending and added to the pending queue via the call to `send_request()` [here](https://github.com/hyperium/hyper/blob/287d712483aec6671427438d60ed2a72f856fd9f/src/proto/h2/client.rs#L277)
This `poll()` here calls [this](https://github.com/hyperium/h2/blob/fcbef502f41cf05c06f1d7b4939fa288d13f7fbc/src/proto/streams/stream.rs#L324), storing a waker in `self.send_task`. For the next iteration of the loop the stream is marked as pending so [this](https://github.com/hyperium/hyper/blob/287d712483aec6671427438d60ed2a72f856fd9f/src/proto/h2/client.rs#L227) will return `Poll::Pending` and immediately overwrite the above waker.

**Comment by seanmonstar** [comment]
Thank you **so** much for debugging this. Once you feel comfortable the patch is no longer blind, I'd be happy to merge a PR. (Mega bonus points if it becomes clear how to trigger this condition in a unit test, but not required.)

**Comment by jfourie1** [comment]
I spent some more time on this today and would definitely not recommend applying the previous "patch". With that patch applied the request that would have gotten stuck is effectively busy polled until it completes, which is not ideal. The real problem/bug is that [this function](https://github.com/hyperium/h2/blob/fcbef502f41cf05c06f1d7b4939fa288d13f7fbc/src/client.rs#L356), specifically the call to `self.inner.poll_pending_open()`  does not allow more than one waiter. In addition this single waiter is shared with [this](https://github.com/hyperium/h2/blob/fcbef502f41cf05c06f1d7b4939fa288d13f7fbc/src/proto/streams/send.rs#L315) and [this](https://github.com/hyperium/h2/blob/fcbef502f41cf05c06f1d7b4939fa288d13f7fbc/src/proto/streams/send.rs#L353). So the current model of calling `poll_ready()` for multiple SendRequest clones before calling `send_request()` will cause this problem. I'll try to add some unit tests to illustrate the problem. It seems that what is needed is some mechanism to allow multiple concurrent tasks to wait for send stream capacity. One thought I had was to use a semaphore and change [poll_ready()](https://github.com/hyperium/h2/blob/fcbef502f41cf05c06f1d7b4939fa288d13f7fbc/src/client.rs#L356) to poll this semaphore. This semaphore will be released once the request is done and the number of permits allowed on the semaphore will be set to max_send_streams. Does this make sense? Any other ideas?

**Comment by jfourie1** [comment]
I created a pull request for this issue, please see [here](https://github.com/hyperium/hyper/pull/3041). The actual issue turned out to be slightly different than what I originally thought the problem was. Please see the commit log in the PR for further detail. The actual issue is a race condition in ClientTask::poll() and not a problem in the h2 crate.

**Comment by ibotty** [comment]
I don't quiet understand.  AFAICT #3041 is merged.  Doesn't this fix the issue?

**Comment by jeromegn** [comment]
I believe it can be closed.

**Comment by kundu-subhajit** [comment]
@jeromegn, @seanmonstar, @jfourie1 : Please guide us, how to reopen the issue?

**Comment by seanmonstar** [comment]
I'd recommend opening a new issue, with whatever details you can provide. Let's leave this one alone.
